### ISSUE TITLE
PoC: using date.parse() to get unique timezone hashes

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [arkenfox/TZP#18]
snip

**Comment by Thorin-Oakenpants** [comment]
So I did some tests by flipping my windows 7 timezone (I do not have 425 timezones to pick from), and closing/restarting FF and retesting. I also changed my date/time a couple of times and spot checked: and that doesn't affect anything.

Clearly I do not want to test this way to get a minimal set of years to return 425 unique hashes, and I can't even set all 425 timezones. Getting 435 unique results should be possible: maybe I need to parse the IANA dataset to pull out date changes after 1970, and even before (again, that Detroit example springs to mind)

```
	/* tests
			one year used: 2020
			OS date: 26th Oct 2020, 3rd July, 20th April

		    moscow: 3121ee1655d676e2d6fc114df302bd8e07d02beb, 23 matches incl. Europe/Moscow
		kiritimati: 888a5ce5fa87ae2b2c86d796d77ed8dd6ca4c1c1,  1 match    =    Pacific/Kiritimati
         dhaka: 57a8cce9a9f867de508dfde0d76b12ddecca5a21,  9 matches incl. Asia/Dhaka
				 seoul: 2de244bdc1f14d51ce09a2b1a16e45a8c2e5b8d5,  9 matches incl. Asia/Seoul
  pacific time: 0d3b5fae482ac70577007fd6ad2bd4b4235617b7,  5 matches incl. America/Los_Angeles
 mountain time: 32cc25fc67ae5110486ff25bf6765ed4945b0ef5,  7 matches incl. America/Denver

	*/
```

**Comment by kkapsner** [comment]
In Europe DST should be abandoned in a few years (the EU decided to get rid of them but the individual countries have to implement this in law) - so you will have many datapoints in the future.

PS: in the console I see my correct time zone.

**Comment by Thorin-Oakenpants** [comment]
Ahhh .. but I'm testing set dates, and ones in the past (well 2020 and before) so I have diffs that won't change. I can't rely on future changes to the IANA dataset: while that might help entropy, it doesn't help me return a timezone

What this test is for, is to determine your timezone only by looking at the timezone offsets (i.e if you lie about it, and workers are blocked, but you don't spoof Date.parse(). It just would have been nice to code it to run through all the 425 "timezones" and then I could start the array of years etc

i.e I would loop timezones, and for each timezone loop the years, each year being four dates => a single hash per timezone. If a year results !== `0,0,0,0` then it's not your timezone (i.e I would assume it is by default and set it false on the first diff) - if all hashes are unique, only one timezone will match you.

I would check the numbers of unique hashes and timezones per hash etc, and keep messing around until I found a minimal set of years. I wonder if it's possible to use a temporal API polyfill?

**Comment by Thorin-Oakenpants** [comment]
@kkapsner @abrahamjuliot try it now: [clicky](https://arkenfox.github.io/TZP/tests/dst.html)

So pass one I am using just two dates (Jan 1 2020 and July 1 2020) and I loop through the 465 timezones to create matches: this will be a smaller list

Then on pass two, I do it again, but I use the smaller list of matches, but a much larger range of years (still using two days in the year) - this seems to be almost sufficient in getting the timezone: there are some exceptions: e.g. RFP returns a bunch, New Zealand returns Antartica/McMurdo as well.

It's not super fast, that's for sure. Depends. e.g. New Zealand is about 400ms, Some European countries take 1.5s.

There's room for improvement. I downloaded the code and data from IANA, and I think I might be a able to craft a set of dates from that, rather than build one on the fly. `build_dates()` function I started with 1970, but it doesn't catch enough, so I went to 1950 and that seemed alright, until I tried Berlin .. so it's currently at 1940

Anyway .. have a play flipping your timezone and restarting Firefox and retesting

**Comment by Thorin-Oakenpants** [comment]
I did some tweaks (and I can do some more: such as narrowing down the years required, and adding specific dates). While perf is not necessarily the aim here, it ultimately will help. I plan to use this in TZP, not to return a timezone, but to run thought a minimal set of dates once to produce a hash. This test PoC is to check that all timezones return a unique hash: i.e they all end up with a green `match`

🔻 Tweaks
- The first tweak was to only use two days in pass one (I don't think four days in the year adds much given pass two)
   - this gives us a baseline of approx 230-250ms
- The second tweak was to move change pass two's dates to start in 1945 (no one did anything DST changes during WW2, I assume), and to drop 2020 as the last date, since it's redundant
- The third tweak was to add a step to the years generated
   - I chose step 2 (i.e odd years since my start Year is 1945 just so I could nail Berlin just for @kkapsner  *sigh*)
   - this gives us a baseline of about 15-20ms per check in pass two (halved compared to all years)
   - the largest pass one returns are around 30 items: this is OK for this PoC since we can only check one browser TZ at at time
- Future tweaks to pass two
   - if you look at the test results below, you will see four timezones that would be unique if we had added even years, but that doesn't mean _all_ even years, it just means I need to add the even years that cover them
   - likewise, with all odds years (done programmatically), many of those could likely be dropped
   - not high priority, since TZP will only be doing a single test which  is currently around that 15-20ms mark: this will grow as we add more unique dates that are needed, but can shrink as we remove redundant dates

🔻 Testing

Using windows 7, I have 134 timezone options, of which 125 are in the test list of 425. The options in windows are called all sorts of things, I'm referring to what Firefox returned in resolvedOptions

🔻 Thoughts / ideas / notes

- the current `build_dates(1945, 2019, 2)` for pass two is 76 dates, add the two from 2020 = 78 dates total used
- the hit rate of unique matches at ~87% is an excellent base
- need to work out how to test every single timezone
   - windows 10: may have more timezones / use up to date timezone names etc
   - linux: start Firefox with a `TZ` variable
- improve
   - reduce years (surely we don't need every even year)
   - add years (e.g specific ones for e.g. Europe/Kiev - see the 4 listed below)
   - add specific dates for remaining multiple final returns according to IANA
      - some may not be worth it `Antarctica/McMurdo` vs `Pacific/Auckland`

🔻 Summary

```
match
  pass one:  15
  pass two:  89
  pass two:   4 more if we use all years, not just even years
  ---------
   SUBTOTAL 108/425 (87%)
  ---------

found (needs more dates)
  pass two:  16 but not unique
  ---------
   SUBTOTAL  16
  ---------

not found (WTF?)
  pass two:   1 (Asia/Riyadh)
  ---------
   SUBTOTAL   1
  ---------
      TOTAL 125/425 (30%)
  ---------

not handled
  ---------
   SUBTOTAL   9 timezones not in the list
  ---------

```

🔻 Details

<details><summary>click me for details</summary><p>

🔹 pass one: match
- `America/Adak`
- `America/St_Johns`
- `Asia/Kabul`
- `Asia/Kathmandu`
- `Asia/Tehran`
- `Atlantic/Cape_Verde`
- `Australia/Darwin`
- `Australia/Eucla`
- `Australia/Lord_Howe`
- `Pacific/Apia`
- `Pacific/Chatham`
- `Pacific/Easter`
- `Pacific/Kiritimati`
- `Pacific/Marquesas`
- `Pacific/Norfolk`

🔹 pass two: match
- note: "from" groupings !== results are from the same group, it's for how refined the 1st pass is
- from 2
  - `America/Asuncion`
  - `America/Miquelon`
  - `America/Santiago`
  - `Asia/Colombo`
  - `Asia/Kolkata`
  - `Asia/Yangon`
  - `Atlantic/Azores`
- from 3
  - `Pacific/Fiji`
  - `Pacific/Honolulu`
  - `Pacific/Tongatapu`
- from 4
  - `Australia/Hobart`
- from 5 (+85ms)
  - `America/Los_Angeles`
  - `America/Phoenix`
  - `America/Tijuana`
- from 6 (+110ms)
  - `America/Ankorage`
  - `America/Halifax`
- from 9 (+190ms)
  - `America/Denver`
  - `America/Guatemala`
  - `America/Regina`
  - `Asia/Almaty`
  - `Asia/Chita`
  - `Asia/Dhaka`
  - `Asia/Kamchatka`
  - `Asia/Omsk`
  - `Asia/Seoul`
  - `Asia/Tokyo`
  - `Asia/Vladivostok`
  - `Asia/Yakutsk`
  - `Australia/Brisbane`
  - `Pacific/Port_Moresby`
- from 10
  - `America/Cancun`
  - `Asia/Magadan`
  - `Asia/Sakhalin`
  - `Asia/Srednekolymsk`
  - `Pacific/Bougainville`
  - `Pacific/Guadalcanal`
- from 13
  - `Asia/Baku`
  - `Asia/Tbilisi`
  - `Asia/Yerevan`
  - `Europe/Astrakhan`
  - `Europe/Samara`
  - `Europe/Saratov`
  - `Indian/Mauritius`
- from 14
  - `Asia/Barnaul`
  - `Asia/Hovd`
  - `Asia/Karachi`
  - `Asia/Krasnoyarsk`
  - `Asia/Novosibirsk`
  - `Asia/Tomsk`
  - `Asia/Yekaterinburg`
- from 15
  - `Africa/Casablanca`
  - `Asia/Irkutsk`
  - `Asia/Pyongyang`
  - `Asia/Shanghai`
  - `Asia/Taipei`
  - `Asia/Ulaanbaatar`
  - `Australia/Perth`
- from 16 (+330ms)
  - `Africa/Cairo`
  - `Africa/Khartoum`
  - `Africa/Tripoli`
  - `Africa/Windhoek`
  - `America/Chicago`
  - `Atlantic/Reykjavik`
  - `Europe/Chisinau`
  - `Europe/Kaliningrad`
- from 19 (+380ms)
  - `America/Grand_Turk`
  - `America/Havana`
  - `America/Indiana/Indianapolis`
  - `America/Port-au-Prince`
- from 20
  - `Asia/Amman`
  - `Asia/Beirut`
  - `Asia/Damascus`
  - `Asia/Jerusalem`
  - `Europe/Bucharest`
- from 23
  - `Asia/Baghdad`
  - `Europe/Istanbul`
  - `Europe/Minsk`
  - `Europe/Moscow`
- from 27 (+500ms)
  - `America/Araguaina`
  - `America/Argentina/Buenos_Aires`
  - `America/Bahia`
  - `America/Cayenne`
  - `America/Montevideo`
  - `America/Punta_Arenas`
  - `America/Sao_Paulo`
- from 31 (+590ms)
  - `America/Caracas`
- from 33 (+630ms)
  - `Europe/Berlin`
  - `Europe/Budapest`
  - `Europe/Warsaw`

🔹 pass two: matches (if we add even years)
- `America/Mexico_City`
  - America/Matamoros, America/Merida, America/Mexico_City, America/Monterrey (from 16)
- `America/Bogota`
  - America/Bogota, America/Guayaquil (from 10)
- `America/Cuiaba`
  - America/Campo_Grande, America/Cuiaba (from 31)
- `Europe/Kiev`
  - Europe/Kiev, Europe/Zaporozhye (from 20)

🔹 pass two: multiple but still detected
- `America/Chihuahua`
  - America/Chihuahua, America/Ojinaga (from 9)
- `America/New_York`
  - America/New_York, America/Toronto (from 19 = +400ms)
- `America/La_Paz`
  - 15 on second pass if all years used
  - 16 on second pass (from 31)
- `Europe/London`
  - Europe/Guernsey, Europe/Isle_of_Man, Europe/Jersey, Europe/London (from 9)
- `Europe/Paris`
  - Europe/Amsterdam, Europe/Brussels, Europe/Luxembourg, Europe/Monaco, Europe/Paris (from 33)
  - Europe/Monaco, Europe/Paris = if all years
- `Africa/Lagos`
  - 11 results (from 15)
  - 10 results if all years used (America/Tunis drops off)
- `Asia/Hebron`
  - Asia/Gaza, Asia/Hebron (from 20)
- `Africa/Johannesburg`
  - 11 results (from 16)
- `Africa/Nairobi`
  - 10 results (from 23)
- `Asia/Dubai`
  - Asia/Dubai, Asia/Muscat, Indian/Mahe, Indian/Reunion (from 13)
- `Asia/Tashkent`
  - Asia/Dushanbe, Asia/Tashkent (from 14)
- `Asia/Bangkok`
  - Asia/Bangkok, Asia/Phnom_Penh, Asia/Vientiane, Indian/Christmas (from 14)
- `Asia/Singapore`
  - Asia/Kuala_Lumpur, Asia/Singapore (from 15)
- `Australia/Adelaide`
  - Australia/Adelaide, Australia/Broken_Hill (from 2)
- `Australia/Sydney`
  - Australia/Currie, Australia/Melbourne, Australia/Sydney (from 4, Australia/Hobart drops off)
- `Pacific/Auckland`
  - Antarctica/McMurdo, Pacific/Auckland (from 3, Pacific/Fiji drops off)

🔹 no match WTF?
- `Asia/Riyadh`
  - pass 1: is in set of 23 results
  - pass 2: nothing

🔹 strikeout (resolvedOption not in list: double check)
- `Etc/GMT+11`
  - Pacifc/Midway, Pacific/Pago_Pago
- `Etc/GMT+12`
  - got nothing
- `Etc/GMT+9`
  - got Pacific/Gambier
- `Etc/GMT+8`
  - got Pacific/Pitcairn
- `Etc/GMT+2`
  - pass one: America/Noronha, Atlantic/South_Georgia
  - pass two: Atlantic/South_Georgia
- `Etc/GMT-12`
  - pass two: Pacific/Funafuti, Pacific/Tarawa, Pacific/Wake, Pacific/Wallis (from 9 results)
- `Etc/GMT-13`
  - pass one: Pacific/Enderbury, Pacific/Fakaofo, Pacific/Tongatapu
  - pass two: got nothing
- `America/Nuuk`
  - pass one: America/Godthab, America/Miquelon
  - pass two: America/Godthab
- `UTC` (set by OS: same as flipping RFP)
  - pass two: 12 results (from 16)
  - pass two: 11 results doing all years

</p></details>

**Comment by Thorin-Oakenpants** [comment]
Well ... I've made a some major breakthroughs ... starting with 1 year for pass one that combined return 328 different buckets for the 425 timezones. Of these 328 buckets, 305 only have a single timezone in them. The numbers `328` + `305` can drop ever so slightly depending on what your timezone is.

So that leaves approx 120 timezones that aren't unique on the first pass. The numbers of timezones in the remaining buckets in descending order are 22, 14, 10, 10, 10, 8, 7, 4 ... etc. Again, the numbers may vary slightly.

So ~70% of time zones should get resolved on pass one. When we move to pass two if needed, it looks like one other magical year resolves about 80% of those ... which leaves about 25 or so left, and I've already worked some of those out: waiting to get all the results so I can see what years are in common

I've also solved some other mysteries
- why some items don't return any matches on the first pass, like `Asia/Riyadh`
- that some results can't get to a single match: according to IANA, e.g
   - > Use Europe/London for Jersey, Guernsey, and the Isle of Man
   - i.e there are zero dates in the entire universe that can help
- aliases/old names: working on them
   - e.g. if I add `America/Nuuk` to the list, it shows up

So all in all, I have a second PoC which I built to calculate which year creates the most buckets and the number of timezones in each bucket (for pass one). The same PoC can run a combined years (i.e the two magic years and any others I might add) and output the results: to see if we can reach the maximum entropy

so, for example
```
America/Indiana/Indianapolis, add any of 1941 1946-50 1952-1953 1956-1957 1959-61 1971-73.. stopping there
America/Cuiaba, add any of 2001 2003 2004
America/Santiago, add any of 2017-2019
America/Araguaina, add any of 1989-1990 1996-2003 2013
America/Bahia, add any of 1991-1999 2001 2003 2012
```

^^ after my two magic years, if I added 2003 that could also cover Cuiba, Araguaina, Bahia, and maybe Indianapolis. Add something for Santiago, and I'm batting 100% so far on the current PoC of returning a matching timezone: have tested about 60 timezones out of 125 possible on my windows 7. That means pass two is lightning fast

Does anyone care?

I'm wondering if I can run a script in the console with a principal selected (like about:config), and assert the TZ in a multiline test? Is anyone interested?

**Comment by Thorin-Oakenpants** [comment]
@abrahamjuliot ... more posts to follow

- home: https://arkenfox.github.io/TZP/tests/tzyears.html (work in progress: more to add/do)
- purpose
   - tool to check how a year(s) affects entropy in order to get the minimal number of dates (`days` x `years`) for max entropy
- how it works
   - gets each date-times offsets in a) UTC and b) each timezone (so 425 loops) and calculates c) the difference
   - the overall hash (currently only in console) is a hash of the local groups so I can tell if anything changed: i.e it's not about the numerical differences
- in TZP we would simply do it once
    - i.e. for each date compare UTC offset with *your* offset - 18 dates = about 2ms?
- any questions?

🔷  TZYEARS

I added two types here - `combine` and `compare`. There will be an input field added where you can add year delimited values

🔻 `combine`

-  ✅ STABILITY: the overall hash is stable
   - since at least FF60-84
   - doesn't seem to matter what your timezone is (it shouldn't)
   - is identical in chrome/opera
   - may change as Firefox or other browsers update their IANA datasets (e.g. [1673027](https://bugzilla.mozilla.org/show_bug.cgi?id=1673027)) ...  however, the highest year I am using is 2018 so it's notreliant on future dates, and it seems the past is pretty much fixed (see since FF60)

- 🗒️  TODO (among other things)
   - display summary at the top of number of groups returned, number of groups with a single timezone, overall groups hash, years used, days used, etc
   - add an input
   - if input is empty run my example
   - there are `22` groups left where I need to confirm that they can't be further split: e.g. IANA database explicitly says that Europe/Guernsey, Europe/Isle_of_Man, Europe/Jersey, Europe/London are identical. There are I think some differences historically, e.g. where on a certain date **AND TIME** there was a difference such as A changed clocks at midnight but B changed clocks at 4am: all these really weird old-timey anachromisms, and I'm not sure if browsers support those. It's something I put to one side for a possible pass three on the other test
   - there are other timezone names not on the list, such as `UTC`, or some available in windows 7 such as `America/Nuuk`, which I need to investigate and perhaps add. The good news is that (I think), they all end up in an existing group: so it may be a case that we can ignore them as they are all just old aliases but backwards compatible
   - run all years from 1879-1920 inclusive with four days per year to make sure 2 days is sufficent

🔻 `compare`

Pretty simple. I used this to run each year in an array and get that years stats: this is what gave me 1879 as the first date to use as it had the most groups and groups with a single timezone (anything prior to 1879 was the same, but I only went back to 1850)

🔻 `method`

Sorry if this a tad confusing
   - I started with just `1879` in the VAR `base years array`
   - blocks of years are all the same (with 1879), and we only need one from each block
       - then I would PUSH the next year to the `base year array` and run it. If there was no diff, I could discard that year (i.e no longer push it) and instead push the next one and test again - repeat
   - but some blocks already cover earlier blocks (I was ascending tests from 1879)
      - if there was a difference (with say `1975`), I would note the hash and check the next year (e.g. `1976`): if it matched the previous test I could discard one of them. If it didn't match, then I would push both of them (e.g. `base years array` and `1975` and `1976`). If that result was the same as `1976`, that meant anything from `1975` was covered by `1976` and we could drop `1975`. If it was not the same, that meant we would add `1975` permanently to the VAR `base years array` for testing the next year etc - repeat

So...
   - I worked out (with `1879` as a base), that there were eight buckets of years, of which adding any year from each bucket would be do the same job
   - I changed which year I chose a bit as I worked out what I was doing. At the start I added the first year (e.g. `1884` was the first year that caused the next change), then I flipped around the `1921` mark I think and added the last year (e.g. `1976` is the last year in that year bucket that caused a change), and the final year bucket of `1977-2020`, I chose `2018` in case of tzdata updates to browsers over the last few years
   - and then to prove a point, I ran all years from `1879-2020` inclusive in a single test and got the same result as running just my nine minimal years of `[1879,1884,1894,1900,1921,1952,1957,1976,2018,]`
   - I also ran `1879` and maybe a couple of others with 4 days per year (one for reach season as maybe some DSTs changes might fall in between a six month gap), but it didn't change any results. I should test this fully

tl:dr;
- take my set of `[1879,1884,1894,1900,1921,1952,1957,1976,2018,]`
- remove any year and you lose entropy
- add any year and you gain no more entropy

Comments welcome. Next post will be about the other test which I renamed: it is now https://arkenfox.github.io/TZP/tests/tz.html

**Comment by Thorin-Oakenpants** [comment]
> Next post will be about the other test 

Actually, this post is about using the multi-line console to bulk test browser set TZ variables. If I open a principal page (like about:config) and then the browser console, I can assert a TZ, run a test and debug the contents, assert the next TZ and repeat

This allows a few things
- on windows (win7 at least) there are only 125 timezones I can set: plus FF has to be restarted
- on linux I can set a FF TZ variable and the start FF (again, so many close/restarts)
- for the `TZYEARS` test, it would mean an easy way to bulk test (in small chunks) that not only is the result stable over releases and cross browser, but also timezone resistant (my logic says it **must* be, but it never hurts to test things)
- for the `TZ` test, where I'm seeing if I can guess the actual timezone name, it allows me to identify on which browser TZ that the real timezone (from resolvedOptions, with no spoofing) is not in pass one or pass two - and the reason for that is the next post

**Comment by Thorin-Oakenpants** [comment]
So this post is about the second test

home: https://arkenfox.github.io/TZP/tests/tz.html (again, a work in progress)
purpose: to see if it would be possible to get the actual name of a user's timezone using just dates
reality:
   - it would have been nice to surprise users who used extensions to spoof it with an exact match
   - pretty sure we can't always get a match (about 20% of the 425 so far)
   - it's too slow to use on TZP (because we have to loop at least 415 timezones)
   - we already have the other test which also gives max entropy
   - TZP only needs to test once... that is UTC vs YOUR timezone (by not specifying any timeZone value) = 1 or 2ms or something

🔻 how it works:

- gets each date-time's offsets in a) UTC and b) each timezone and calculates c) the difference
- as it loops thru each timezone, if the combined result of all the dates are all zero's then it's considered a potential match
- does passes
   - pass one is a single year (two days per year): this is a minimum 230 odd ms, you can't reduce that
   - if pass one returns a single result, then that must be the match, otherwise we try pass two (assuming we didn't get returned absolutely nothing)
   - pass two now uses a vastly smaller list of timezones to check. Using `1879` as our pass one year, the largest possible second pass is `22` items, and the next ones after that are `14`, then `10`
   - so pass two can now use our other eight years of `1884,1894,1900,1921,1952,1957,1976,2018` = awesome perf, and we should get max entropy: i.e we probably can't drill the potential matches down any further

🔻 but there's issues that make the real timezone not get matched

Using `1879`, 1879 (being old-timey with weird anachronisms), gives some nutty results. e.g
   - when you've set the FF TZ to `Europe/Paris`, when you test `Europe/Paris` from the list of 425 timezones, instead of returning `0, 0` it returns `530.65, 530.65`
    - Africa/Lagos returns `766.4, 766.4`
    - obviously, dropping the real match means we fail and it does not end up in the next pass
    - so being able to test every single FF TZ variable and check which timezones are not matched in both pass one and pass two would be a way to hardcode workarounds for them

The same happens on some years, there are tiny diffs (leap seconds etc), e.g.
   - with FF TZ set as `X` when you test `X` as it loops thru the list of timezones, it does not returns zeros
   - 1921 `Amercia/Chihuahua` is `0.3333333333333333, 0.3333333333333333`
   - 1919 `Amercia/Mexico_City` is 0.6, 0.6

So I added some flooring
```js
	if (number < 1 && number > -1) {number = 0}
```

then I hit Berlin which returned `-1, -1` (I didn't note the year), so I used
```js
	if (number < 1.01 && number > -1.01) {number = 0}
```

but, and I have yet to do more on this, I think that then messed up the earlier ones, like `Chihuahua` and `Mexico_City`, so I started doing this
```js
	// special flooring
	let isInclusive = false
	if (tz == "Europe/Berlin") {isInclusive = true} // -1, -1
	// floor
	if (isInclusive) {
		let before = number
		if (number < 1.01 && number > -1.01) {number = 0}
		let after = number
		if (before != after) {
			console.debug("special flooring", tz, before, "=>", number)
		}
	} else {
		if (number < 1 && number > -1) {number = 0}
	}
```

So that's about where I'm at
 - need to look at old aliases names like `America/Nuuk`
 - need to check no more entropy can be gained
 - confirm stability if FF TZ changes
 - confirm 4 days per year doesn't add any entropy
 - confirm that groups of more than one timezone can't be reduced with one-off special date-times
 - add test to TZP

And for the get your real timezone name test which is not for TZP but it'd be nice to make it as robust as possible
- programmatically set FF TZ, test TZ is in pass one and pass two (using our minimum nine years), and if not, debug the values
- test if I just always floor isInclusive to -1 to 1, or split it
- hardcode exceptions for anachromisms over 1 or under 1?
- or change the floor values (e.g. what if I get 1.33333, or 2.5)

That's my story... hope you found it interesting. Sorry that  https://arkenfox.github.io/TZP/tests/tz.html is still using the old crap, 2020 for pass one and then 1945 + odd years on pass two: and no flooring.

My new PoC is miles better and faster. I just want to fix up those outstanding weird value issues first

Comments welcome

**Comment by Thorin-Oakenpants** [comment]
sorry: to clarify

I use date.parse() to get timezone offsets between two dates and calculate the difference
- TZYEARS
   - uses a `UTC` date vs the same date in `each timezone` ... no OS/FF variables involved (hence why it's stable)
- TZ 
   - uses the same but I'm tackling it differently to make as fast as possible
   - maybe uses a `UTC` date vs the same date as `you` (no timeZone value set) .. thus if the diff is zeros it's a potential match
- TZP
   - when I add the poc
   - uses a `UTC` date vs `you` (i.e no timezone value applied) ... a single run in 2ms or whatever for a hash

So I've just clicked... I can tweak my code with a hardcoded (or exposed) boolean to mix that up
- i.e in TZYEARS  i should calculate diffs
  - use `a specific timezone` (instead of UTC) vs the same date in `each timezone`
  - so no OS/FF variables need setting
  - however this is probably moot since it already doesn't rely on any OS settings (but I could test it anyway)
- i.e in TZ I think I can simulate the same thing

So in fact I can simulate all I need without asserting with special priviledges

This also opens up another avenue, when I can store a hash for each of the 425 timezones (store it in the timezone array), and just do the 2ms test to return the result by directly looking it up - might still do that as a secondary test PoC, as I still like the drilling down in two passes (or more)

**Comment by kkapsner** [comment]
You gave me a very good point why I should not include timezone spoofing into CB. It's just a giant mess.

**Comment by Thorin-Oakenpants** [comment]
> It's just a giant mess

I only got about 8 hrs sleep over the last 2.5 days, and ended up confusing myself ~~a little~~ quite a bit, but I've worked out some more stuff now I've slept, but yeah .. it's just a giant mess and combined with a lack of sleep does my head in

PS: I ran tzyears with4 days and all years 1879-2020 inclusive and got the same hash, so no extra entropy using Jan 1 and July 1.. and I guess any two dates 6 months apart. If I actually put my thinking hat on, if there are DST changes, it changes twice a year- where that change falls doesn't matter as the two values combined will be 12 months (e.g. a short DST of 4 months and a non-DST of 8 months - there will always be a diff between two equal sizes of 6 months)

... so yeah, 4 days is moot (assuming my thinking hat isn't damaged)

**Comment by Thorin-Oakenpants** [comment]
Alrighty... I've had some extra strong coffee and a break and made some decisions

The tz test where I do two passes is redundant. I prototype, and that was the first page.. then I wanted something to workout the best years to use, and that's the tzyears one. Both tests do the nine date, so we only need one. The tzyears one actually shows the values, and there is zero need to deal with any weird numbers: I only need as unique a hash as possible for TZP, not to always name the timezone itself (which is impossible given some are always in a small group)

So I don't care about making almost zeros into zeros, or dealing with weird old-timey anachromisms.

So tzyears it is, I'll be adding some extra tweaks: While I know my set of years/days now, I still want to add in UTC and other older/aliases

---

I'm not even sure how I did this: but the tzyears test was designed to be stable: i.e I use an array of set dates. I specify UTC for one and a timezone for the other - i.e I wasn't trying to use the current timezone at all. While this means the entropy (what locales are in what groups etc) is stable regardless of your own timezone: and I've worked out the max enbtropy!

I though that would also make the differences in the offsets the same. But they change: but they all change, (and by the same amount I think although it's probably relative to each other, some will change more depending on DST). But I just don't get it: and yet that's exactly what I was exploiting in the first test by detecting all zeros. I'm missing something as to why/how the local timezone is changing values - but TBH, it's not important except to satisfy my curiosity

Anyway: I did some more tests
- I tested (in large chunks of 200 years) all years from the year 0 to 1878 with my set of nine years, and for good measure I also threw in 300 odd random years BC going back to 1500BC or something .. zero extra entropy
- I also played with checking all differences => integers
   - the reason was that there **LOTS** of timezones that have decimals (just look at the tzyears test), so they'll never be all zeros: if I did this then only weird outliers like that Paris one wouldn't get caught, and if made all values into integers, then the entropy wouldn't change
   - it tried three ways: stripping the decimal (so it's always closer to zero), floor, and ceil
   - I did a handful of spot tests between OS timezone changes
   - but alas, **the entropy changes**, and the entropy can differ depending on your timezone: and it can differ between the methods per timezone (but not always)
   - it didn't seem to drop much: lowest I got was five less groups (current max is 348) and eight less unique timezones (current max is 326). The things that changed were things like
   // America/Argentina/Salta        => joined America/Argentina/Jujuy
   // America/North_Dakota/New_Salem => joined America/North_Dakota/Center
   // Indian/Mahe    => joined "Asia/Dubai,Asia/Muscat"
   // Indian/Reunion => joined "Asia/Dubai,Asia/Muscat"

```js
	//force integers
	if (Number.isInteger(diff) == false) {
		// option 1: strip decimals
		if (diff > 0) {
			diff = Math.floor(diff)
		} else {
			diff = Math.ceil(diff)
		}
		// option 2: always floor
			//diff = Math.floor(diff)
		// option 2: always ceil
			//diff = Math.ceil(diff)
	}
```

Thought I would share. Makes sense that it's not stable, since they all seem to move +/- 26 hrs relative to the spread of possible timezones anyway, trying to make into integers between negatives/positives is too hard to do

For example: here's Berlin 2018 at two extreme timezones
- `Etc/GMT+12` is the resolved option, windows OS options calls it `UTC -12:00`
   - `780, 720`
- `PacificKiritimati`, windows OS calls this (UTC +14:00)
   - `-780,  -720`

So the max spread is 26 hrs (diff between `780` and `-780`): which means almost everything (I guess you call always exclude the two extreme ends) can change between positive/negative. Depending on the decimal value, stripping decimals, flooring, and ceil all seem to cause changes to entropy as values move between positive/negative

(I still can't wrap my head around why the values change)

That said, when running multiple passes and using integers on almost zeros wouldn't affect entropy, because we're drilling down: and only one group will match - but I'm not bothering to do that anymore

/sorry for the long post.. I wonder if anyone finds this interesting

**Comment by Thorin-Oakenpants** [comment]
OK, am onto it now. You're always detected if your timezone is on the list.

If you look at the test results, the 349 (I added `UTC` as a timezone), almost all of them have decimals. This means they will never be zero. But since I'm running the whole thing (all years) and not in multiple passes - I do not need to check for all zeros, as I can just generate the hash *similar* to how I would on TZP and match it from there

❓ goddamn it, why does it take so long sometimes for the light-bulb to click on when prototyping and learning

So there you go. I can now return the real timezone name in 327 cases, and a small group in the other 99 cases - it's live BTW, enjoy (still have more things to add, BTW)

Here's New York
![bout-F-time](https://user-images.githubusercontent.com/16656956/97811534-22776980-1c73-11eb-8941-589d864094c3.png)

**Comment by Thorin-Oakenpants** [comment]
FYI: the multi-pass test is gone. The one and only test is now at https://arkenfox.github.io/TZP/tests/timezones.html .. and it's not going to move

todo:
- add more misc timezones (currently UTC)
- work out how to color unique timezone in green now we merge two lists (misc and iana, which have different rules for parsing): the combined list order is different to the legend order (it currently works except for UTC, because UTC is alphabetically the last entry)
- add an input field: if empty run my example, if not empty, parse for comma delimited years
- add intro
- add to tzp index.html

PS: added a veracity output for kicks .. enjoy
